File: src/bmsx/render/vdp/framebuffer.ts
import type { TextureHandle } from '../backend/backend';
import type { VDP } from '../../machine/devices/vdp/vdp';
import type { VdpFrameBufferPresentation, VdpFrameBufferPresentationSink } from '../../machine/devices/vdp/device_output';
import { FRAMEBUFFER_RENDER_TEXTURE_KEY, FRAMEBUFFER_TEXTURE_KEY } from '../../rompack/format';
import { DEFAULT_TEXTURE_PARAMS } from '../backend/texture_params';
import type { TextureManager } from '../texture_manager';
import type { GameView } from '../gameview';

const EMPTY_TEXTURE_SEED = new Uint8Array(4);
const VDP_FRAMEBUFFER_TEXTURE_PARAMS = Object.freeze({
	...DEFAULT_TEXTURE_PARAMS,
	srgb: false,
});

export class VdpFrameBufferTextures implements VdpFrameBufferPresentationSink {
	private renderFrameBufferTexture: TextureHandle = null as TextureHandle;
	private displayFrameBufferTexture: TextureHandle = null as TextureHandle;
	private frameBufferTextureWidth = 0;
	private frameBufferTextureHeight = 0;

	public constructor(
		private readonly textureManager: TextureManager,
		private readonly view: GameView,
	) {
	}

	public consumeVdpFrameBufferPresentation(presentation: VdpFrameBufferPresentation): void {
		const presentationCount = presentation.presentationCount;
		for (let index = 0; index < presentationCount; index += 1) {
			this.presentVdpFrameBufferPages();
		}
		const frameBufferWidth = presentation.width;
		const frameBufferHeight = presentation.height;
		this.frameBufferTextureWidth = frameBufferWidth;
		this.frameBufferTextureHeight = frameBufferHeight;
		if (!presentation.readbackValid) {
			return;
		}
		if (presentationCount !== 1 || presentation.requiresFullSync) {
			this.view.backend.updateTextureRegion(
				this.textureManager.getTextureByUri(FRAMEBUFFER_RENDER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
				presentation.renderReadback,
				frameBufferWidth,
				frameBufferHeight,
				0,
				0,
				VDP_FRAMEBUFFER_TEXTURE_PARAMS
			);
			this.view.backend.updateTextureRegion(
				this.textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
				presentation.displayReadback,
				frameBufferWidth,
				frameBufferHeight,
				0,
				0,
				VDP_FRAMEBUFFER_TEXTURE_PARAMS
			);
			return;
		}
		const rowBytes = frameBufferWidth * 4;
		const displayReadback = presentation.displayReadback;
		const spans = presentation.dirtySpansByRow;
		for (let row = presentation.dirtyRowStart; row < presentation.dirtyRowEnd; row += 1) {
			const span = spans[row]!;
			if (span.xStart >= span.xEnd) {
				continue;
			}
			const byteStart = row * rowBytes + span.xStart * 4;
			this.view.backend.updateTextureRegion(
				this.textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
				displayReadback,
				span.xEnd - span.xStart,
				1,
				span.xStart,
				row,
				VDP_FRAMEBUFFER_TEXTURE_PARAMS,
				byteStart
			);
		}
	}

	public initialize(vdp: VDP): void {
		this.frameBufferTextureWidth = vdp.frameBufferWidth;
		this.frameBufferTextureHeight = vdp.frameBufferHeight;
		this.renderFrameBufferTexture = this.textureManager.createTextureFromPixelsSync(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			EMPTY_TEXTURE_SEED,
			1,
			1,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.renderFrameBufferTexture = this.textureManager.resizeTextureForKey(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = this.renderFrameBufferTexture;
		this.displayFrameBufferTexture = this.textureManager.createTextureFromPixelsSync(
			FRAMEBUFFER_TEXTURE_KEY,
			EMPTY_TEXTURE_SEED,
			1,
			1,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.displayFrameBufferTexture = this.textureManager.resizeTextureForKey(
			FRAMEBUFFER_TEXTURE_KEY,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_TEXTURE_KEY] = this.displayFrameBufferTexture;
		vdp.syncFrameBufferPresentation(this);
	}

	public width(): number {
		return this.frameBufferTextureWidth;
	}

	public height(): number {
		return this.frameBufferTextureHeight;
	}

	public displayTexture(): TextureHandle {
		return this.displayFrameBufferTexture;
	}

	public renderTexture(): TextureHandle {
		return this.renderFrameBufferTexture;
	}

	private presentVdpFrameBufferPages(): void {
		this.textureManager.swapTextureHandlesByUri(
			FRAMEBUFFER_TEXTURE_KEY,
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_TEXTURE_KEY] = this.textureManager.getTextureByUri(
			FRAMEBUFFER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = this.textureManager.getTextureByUri(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		const renderTexture = this.renderFrameBufferTexture;
		this.renderFrameBufferTexture = this.displayFrameBufferTexture;
		this.displayFrameBufferTexture = renderTexture;
	}
}

File: src/bmsx_cpp/render/vdp/framebuffer.cpp
#include "render/vdp/framebuffer.h"

#include "machine/devices/vdp/contracts.h"
#include "machine/devices/vdp/vdp.h"
#include "render/gameview.h"
#include "render/texture_manager.h"
#include "rompack/format.h"
#include <array>
#include <utility>
#include <vector>

namespace bmsx {
namespace {

const std::array<u8, 4> EMPTY_TEXTURE_SEED{{0, 0, 0, 0}};
const TextureParams VDP_FRAMEBUFFER_TEXTURE_PARAMS{
	.size = {.x = 0.0f, .y = 0.0f},
	.wrapS = TEXTURE_WRAP_CLAMP_TO_EDGE,
	.wrapT = TEXTURE_WRAP_CLAMP_TO_EDGE,
	.minFilter = TEXTURE_FILTER_NEAREST,
	.magFilter = TEXTURE_FILTER_NEAREST,
	.srgb = false,
};

} // namespace

VdpFrameBufferTextures::VdpFrameBufferTextures(TextureManager& textureManager, GameView& view)
	: m_textureManager(textureManager)
	, m_view(view) {
}

void VdpFrameBufferTextures::consumeVdpFrameBufferPresentation(const VdpFrameBufferPresentation& presentation) {
	presentVdpFrameBufferPages(presentation.presentationCount);
	const u32 width = presentation.width;
	const u32 height = presentation.height;
	m_frameBufferTextureWidth = width;
	m_frameBufferTextureHeight = height;
	if (!presentation.readbackValid) {
		return;
	}
	if (presentation.presentationCount != 1u || presentation.requiresFullSync) {
		m_view.backend()->updateTextureRegion(
			m_textureManager.getTextureByUri(FRAMEBUFFER_RENDER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
			presentation.renderReadback->data(),
			static_cast<i32>(width),
			static_cast<i32>(height),
			0,
			0,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		m_view.backend()->updateTextureRegion(
			m_textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
			presentation.displayReadback->data(),
			static_cast<i32>(width),
			static_cast<i32>(height),
			0,
			0,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		return;
	}
	const u32 rowBytes = width * 4u;
	const auto& pixels = *presentation.displayReadback;
	const auto& spans = *presentation.dirtySpansByRow;
	for (u32 row = presentation.dirtyRowStart; row < presentation.dirtyRowEnd; ++row) {
		const auto& span = spans[row];
		if (span.xStart >= span.xEnd) {
			continue;
		}
		const size_t byteOffset = static_cast<size_t>(row) * static_cast<size_t>(rowBytes) + static_cast<size_t>(span.xStart) * 4u;
		m_view.backend()->updateTextureRegion(
			m_textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, VDP_FRAMEBUFFER_TEXTURE_PARAMS),
			pixels.data() + byteOffset,
			static_cast<i32>(span.xEnd - span.xStart),
			1,
			static_cast<i32>(span.xStart),
			static_cast<i32>(row),
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
	}
}

void VdpFrameBufferTextures::initialize(VDP& vdp) {
	m_frameBufferTextureWidth = vdp.frameBufferWidth();
	m_frameBufferTextureHeight = vdp.frameBufferHeight();
	m_renderFrameBufferTexture = m_textureManager.createTextureFromPixelsSync(
		FRAMEBUFFER_RENDER_TEXTURE_KEY,
		EMPTY_TEXTURE_SEED.data(),
		1,
		1,
		VDP_FRAMEBUFFER_TEXTURE_PARAMS
	);
	m_renderFrameBufferTexture = m_textureManager.resizeTextureForKey(
		FRAMEBUFFER_RENDER_TEXTURE_KEY,
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		VDP_FRAMEBUFFER_TEXTURE_PARAMS
	);
	m_view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = m_renderFrameBufferTexture;
	m_displayFrameBufferTexture = m_textureManager.createTextureFromPixelsSync(
		FRAMEBUFFER_TEXTURE_KEY,
		EMPTY_TEXTURE_SEED.data(),
		1,
		1,
		VDP_FRAMEBUFFER_TEXTURE_PARAMS
	);
	m_displayFrameBufferTexture = m_textureManager.resizeTextureForKey(
		FRAMEBUFFER_TEXTURE_KEY,
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		VDP_FRAMEBUFFER_TEXTURE_PARAMS
	);
	m_view.textures[FRAMEBUFFER_TEXTURE_KEY] = m_displayFrameBufferTexture;
	vdp.syncFrameBufferPresentation(*this);
}

void VdpFrameBufferTextures::presentVdpFrameBufferPages(u32 presentationCount) {
	for (u32 index = 0; index < presentationCount; ++index) {
		m_textureManager.swapTextureHandlesByUri(
			FRAMEBUFFER_TEXTURE_KEY,
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		m_view.textures[FRAMEBUFFER_TEXTURE_KEY] = m_textureManager.getTextureByUri(
			FRAMEBUFFER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		m_view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = m_textureManager.getTextureByUri(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			VDP_FRAMEBUFFER_TEXTURE_PARAMS
		);
		std::swap(m_renderFrameBufferTexture, m_displayFrameBufferTexture);
	}
}

} // namespace bmsx
